handler_test.go

```diff
@@ -0,0 +1,1982 @@
+								Reason:  dns.RcodeToString[dns.RcodeSuccess],
+								Message: rcodeMessage[dns.RcodeSuccess],
+							},
+						},
+					},
+				},
+			},
+		},
+	},
+	{
+		name: "Update regular dns name resolver object status with latest ttl of the corresponding IPs, set resolutionFailures to 0 and Degraded condition to false",
+		dnsNameResolvers: []ocpnetworkapiv1alpha1.DNSNameResolver{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "regular",
+					Namespace: "dns",
+				},
+				Spec: ocpnetworkapiv1alpha1.DNSNameResolverSpec{
+					Name: "www.example.com.",
+				},
+				Status: ocpnetworkapiv1alpha1.DNSNameResolverStatus{
+					ResolvedNames: []ocpnetworkapiv1alpha1.DNSNameResolverResolvedName{
+						{
+							DNSName: "www.example.com.",
+							ResolvedAddresses: []ocpnetworkapiv1alpha1.DNSNameResolverResolvedAddress{
+								{
+									IP:         "1.1.1.1",
+									TTLSeconds: 5,
+								},
+								{
+									IP:         "1.1.1.2",
+									TTLSeconds: 5,
+								},
+							},
+							ResolutionFailures: 2,
+							Conditions: []metav1.Condition{
+								{
+									Type:    ConditionDegraded,
+									Status:  metav1.ConditionTrue,
+									Reason:  dns.RcodeToString[dns.RcodeServerFailure],
+									Message: rcodeMessage[dns.RcodeServerFailure],
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		dnsQueryTestCases: []test.Case{
+			{
+				Qname: "www.example.com.",
+				Qtype: dns.TypeA,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.A("www.example.com. 30 IN A 1.1.1.1"),
+					test.A("www.example.com. 30 IN A 1.1.1.2"),
+				},
+			},
+		},
+		expectedStatuses: []ocpnetworkapiv1alpha1.DNSNameResolverStatus{
+			{
+				ResolvedNames: []ocpnetworkapiv1alpha1.DNSNameResolverResolvedName{
+					{
+						DNSName: "www.example.com.",
+						ResolvedAddresses: []ocpnetworkapiv1alpha1.DNSNameResolverResolvedAddress{
+							{
+								IP:         "1.1.1.1",
+								TTLSeconds: 30,
+							},
+							{
+								IP:         "1.1.1.2",
+								TTLSeconds: 30,
+							},
+						},
+						ResolutionFailures: 0,
+						Conditions: []metav1.Condition{
+							{
+								Type:    ConditionDegraded,
+								Status:  metav1.ConditionFalse,
+								Reason:  dns.RcodeToString[dns.RcodeSuccess],
+								Message: rcodeMessage[dns.RcodeSuccess],
+							},
+						},
+					},
+				},
+			},
+		},
+	},
+	// Regular DNS Name resolution failure
+	{
+		name: "Don't update regular dns name resolver object status",
+		dnsNameResolvers: []ocpnetworkapiv1alpha1.DNSNameResolver{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "regular",
+					Namespace: "dns",
+				},
+				Spec: ocpnetworkapiv1alpha1.DNSNameResolverSpec{
+					Name: "www.example.com.",
+				},
+			},
+		},
+		dnsQueryTestCases: []test.Case{
+			{
+				Qname: "www.example.com.",
+				Qtype: dns.TypeA,
+				Rcode: dns.RcodeNameError,
+			},
+		},
+		expectedStatuses: []ocpnetworkapiv1alpha1.DNSNameResolverStatus{{}},
```

Comment:
The `ResolutionFailures` field is used to keep a count on the number of consecutive failures after a successful lookup. If the `ResolutionFailures` goes beyond 5 then the details of the corresponding DNS name is removed from the status of the DNSNameResolver object. This is mainly for the DNSNameResolver object for a wildcard DNS name, however this also applies to the DNSNameResolver object for a regular DNS name as well.

If the status does not have any resolved names, then it can be either that a successful DNS name lookup hasn't matched the DNSNameResolver object, or the resolved name details were removed due to 5 consecutive unsuccessful lookup. If we add the failure details even when there's no entry for the DNS name in the status, then we might spend time in constantly updating the DNSNameResolver status for the failure and then removing it, and then again adding it back. We are trying to remove the stale entries from the status by this process (this is done by the DNSNameResolver controller).